### PR TITLE
GRR Hunt/Flow with osquery

### LINKED ISSUE #512: GRR Hunt using Osquery
Add a GRR Hunt collector that runs an osquery flow.

### PR DESCRIPTION
**Pull request by sydp**
Fixes #512.

### CODE REVIEW

File: tests/lib/collectors/grr_hosts.py [hunk 1]
```diff
@@ -585,6 +588,67 @@ def testProcessFromContainers(self,
...
+      self.grr_osquery_collector.Process(container)
+    self.grr_osquery_collector.PostProcess()
+
+    mock_LaunchFlow.assert_called_with(
+        mock_grr_hosts.MOCK_CLIENT_RECENT,
+        'OsqueryFlow',
+        osquery_pb2.OsqueryFlowArgs(
+            query='SELECT * FROM processes',
+            timeout_millis=
+                grr_hosts.GRROsqueryCollector.DEFAULT_OSQUERY_TIMEOUT_MILLIS,
```
Comment (tomchop) [inline tests/lib/collectors/grr_hosts.py:644]: What's the expected value here? Please always favor the use of litterals in tests, as it makes it easier to understand what the test is actually testing for.

**Review by tomchop — CHANGES_REQUESTED** [review]

File: tests/lib/collectors/grr_hosts.py (hunk 1 above)
Comment (sydp) [reply tests/lib/collectors/grr_hosts.py]: Done.

File: dftimewolf/lib/collectors/osquery.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.logger.warning('Path {0:s} does not exist'.format(path))
+        continue
+
+      with open(path, mode='r') as fd:
+        for line_number, line in enumerate(fd.readlines()):
+          if not line.upper().startswith('SELECT '):
+            self.logger.warning('Osquery on line {0:d} of {1:s} does not start '
+                                'with SELECT.'.format(line_number, path))
+          else:
+            osquery_queries.append(self.query)
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/osquery.py]: Done and added test class.

File: dftimewolf/lib/collectors/osquery.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    osquery_queries = []
+    if not self.query.upper().startswith('SELECT '):
+      self.logger.warning(
+          'Osquery parameter does not start with SELECT.')
+    else:
+      osquery_queries.append(self.query)
+
+    for path in self.paths:
+      if not os.path.exists(path):
+        self.logger.warning('Path {0:s} does not exist'.format(path))
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/osquery.py]: Done.

File: dftimewolf/lib/collectors/osquery.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      files (str): osquery filepaths
+    """
+    if not query and not paths:
+      self.ModuleError('Both query and files cannot be empty.', critical=True)
+
+    if query:
+      self.query = query
+
+    if paths:
+      self.paths = [path.strip() for path in paths.split(',')]
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/osquery.py]: Moved most of the action to SetUp as suggested.

File: dftimewolf/lib/collectors/osquery.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            query: str,
+            paths: str) -> None:
+    """Sets up the paths to collect.
+
+    Args:
+      query (str): osquery query.
+      files (str): osquery filepaths
+    """
+    if not query and not paths:
+      self.ModuleError('Both query and files cannot be empty.', critical=True)
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/osquery.py]: Done!

File: dftimewolf/lib/collectors/grr_hunt.py [hunk 6]
```diff
@@ -316,6 +321,89 @@ def Process(self) -> None:
...
+
+  def Process(self) -> None:
+    """Starts a new Osquery GRR hunt."""
+    osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
+
+    for osquery_container in osquery_containers:
+      hunt_args = osquery_flows.OsqueryFlowArgs()
+      hunt_args.query = osquery_container.query
+      hunt_args.timeout_millis = self.timeout_millis
+      hunt_args.ignore_stderr_errors = self.ignore_stderr_errors
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hunt.py:402]: Done (was following the GRR style but have adpted to pass as kwargs.)

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 7]
```diff
@@ -647,6 +649,190 @@ def GetThreadOnContainerType(self) -> Type[interface.AttributeContainer]:
...
+                str(val) for val in (hostname, flow_id, 'csv')))
+
+        with open(output_file_path, mode='w') as fd:
+          container.data_frame.to_csv(fd)
+
+        self.logger.info('Saved {0:s}'.format(output_file_path))
+
+        manifest_fd.write(f'"{flow_id}",')
+        manifest_fd.write(f'"{hostname}",')
+        manifest_fd.write(f'"{query}"\n')
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hosts.py]: No reason anymore.

[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hosts.py]: Fixed.

File: dftimewolf/lib/collectors/grr_hosts.py [hunk 8]
```diff
@@ -647,6 +649,190 @@ def GetThreadOnContainerType(self) -> Type[interface.AttributeContainer]:
...
+    if not self.directory:
+      return
+
+    if not os.path.isdir(self.directory):
+      os.makedirs(self.directory)
+
+    manifest_file_path = os.path.join(self.directory, 'MANIFEST.csv')
+
+    self.logger.info(
+        'Saving osquery flow results to {0:s}'.format(manifest_file_path))
```
[the remark this answers is not in the document]
Comment (sydp) [reply dftimewolf/lib/collectors/grr_hosts.py]: Using f-strings now.

File: tests/lib/collectors/osquery.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    with mock.patch(
+        'builtins.open',
+        new=mock.mock_open(read_data=test_ok_data)) as _:
+      osquery_collector.SetUp(query='', paths='ok')
+
+    osquery_collector.Process()
+
+    containers = test_state.GetContainers(OsqueryQuery)
+    self.assertEqual(len(containers), 1)
```
Comment (tomchop) [inline tests/lib/collectors/osquery.py:87]: Can we test the value of the container query here?

**Review by tomchop — CHANGES_REQUESTED** [review]
One last testing-related nit, otherwise LGTM!

File: tests/lib/collectors/osquery.py (hunk 9 above)
Comment (sydp) [reply tests/lib/collectors/osquery.py:87]: Done.

**Review by tomchop — APPROVED** [review]

**Pull request merged**